src/field/FieldPropagator.i.hh

```diff
@@ -0,0 +1,199 @@
+    : shared_(shared), driver_(driver)
+{
+    CELER_ASSERT(shared_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Propagate a charged particle in a magnetic field and update the field
+ * state. It utilises a magnetic field driver based on an adaptive step
+ * control to track a charged particle until it travels along a curved
+ * trajectory for a given step length within a required accuracy or intersects
+ * with a new volume (geometry limited step).
+ */
+CELER_FUNCTION real_type FieldPropagator::operator()(FieldTrackView* view)
+{
+    CELER_ASSERT(view && view->charge() != zero_quantity());
+
+    real_type step = view->step();
+
+    // If not a valid range, transportation shouild not be a candiate process
+    if (step < shared_.minimum_step)
+    {
+        return numeric_limits<real_type>::infinity();
+    }
+
+    // Initial parameters and states for the field integration
+    real_type    step_taken = 0;
+    OdeState     end_state  = view->state();
+    Intersection intersect;
+
+    do
+    {
+        OdeState  beg_state = end_state;
+        real_type step_left = step - step_taken;
+
+        // Advance within the tolerence error for the remaining step length
+        real_type sub_step = driver_(step_left, &end_state);
+
+        // Check whether this sub-step cross a volume boundary
+        view->on_boundary(intersect.intersected);
+        this->check_intersection(
+            view, beg_state.pos, end_state.pos, &intersect);
+
+        // If it is a geometry limited step, find the intersection point
+        if (intersect.intersected)
+        {
+            intersect.scale *= sub_step;
+            end_state = this->locate_intersection(view, beg_state, &intersect);
+
+            // Update states for field and navigation
+            sub_step      = intersect.step;
+            end_state.pos = intersect.pos;
+            view->on_boundary(intersect.intersected);
+            view->update_vgstates();
+        }
+
+        // Add sub-step until there is no remaining step length
+        step_taken += sub_step;
+
+    } while (!intersect.intersected
+             && (step_taken + shared_.minimum_step) < step);
+
+    // Update the field track view
+    view->state(end_state);
+    view->step(step_taken);
+
+    return step_taken;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Check whether the final position of the field integration for a given step
+ * is inside the current volume or is crossed any boundary of adjacent volumes.
+ */
+CELER_FUNCTION
+void FieldPropagator::check_intersection(FieldTrackView* view,
+                                         const Real3     beg_pos,
+                                         const Real3     end_pos,
+                                         Intersection*   intersect)
+{
+    intersect->intersected = false;
+
+    Real3 chord = end_pos;
```

Comment:
I am not sure but would re-using the argument mem space/register be more efficient (i.e `const Real3 end_pos` -> `Real3 end_pos` and maybe alias it (via a ref) for code readability)?

Reply:
I guess that both **Real3 end_pos** and **const Real3& end_pos** are acceptable, but not **Real3& end_pos.** as it is mutable.  For now, I change it to const Real3& end_pos.